### PR TITLE
[RISCV] Support `llvm.masked.compressstore` intrinsic

### PR DESCRIPTION
**Pull request by nikolaypanchenko**
The changeset enables lowering of `llvm.masked.compressstore(%data, %ptr, %mask)` for RVV for fixed vector type into:
```
%0 = vcompress %data, %mask, %vl
%new_vl = vcpop %mask, %vl
vse %0, %ptr, %1, %new_vl
```
Such lowering is only possible when `%data` fits into available LMULs and otherwise `llvm.masked.compressstore` is scalarized by `ScalarizeMaskedMemIntrin` pass.
Even though RVV spec in the section `15.8` provide alternative sequence for compressstore, use of `vcompress + vcpop` should be a proper canonical form to lower `llvm.masked.compressstore`. If RISC-V target find the sequence from `15.8` better, peephole optimization can transform `vcompress + vcpop` into that sequence.

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/rvv/compressstore.ll [hunk 1]
```diff
@@ -0,0 +1,17545 @@
...
+; RV32-NEXT:    addi a5, a5, -1920
+; RV32-NEXT:    add a5, sp, a5
+; RV32-NEXT:    vsetvli zero, a4, e8, m8, ta, ma
+; RV32-NEXT:    vse8.v v24, (a5)
+; RV32-NEXT:    lbu a4, 1905(a2)
+; RV32-NEXT:    addi a5, a0, 1
+; RV32-NEXT:    sb a4, 0(a0)
+; RV32-NEXT:    mv a0, a5
+; RV32-NEXT:    slli a4, a3, 15
+; RV32-NEXT:    bltz a4, .LBB
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/rvv/compressstore.ll:8607]: Is this codegen right? It's so huge.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/compressstore.ll:8607]: Looks like there's no alignment information so the scalar stores got split

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -1609,3 +1609,10 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
                   C2.NumIVMuls, C2.NumBaseAdds,
                   C2.ScaleCost, C2.ImmCost, C2.SetupCost);
 }
+
+bool RISCVTTIImpl::isLegalMaskedCompressStore(Type *DataTy) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1613]: This interface needs an alignment parameter doesn't it?

Comment (nikolaypanchenko) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1613]: For RVV it's not required. The store of the compressed data is unaligned, i.e. it should be possible to use `vse8` for non-naturally aligned memory

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1613]: We would need to scale the VL from the vcpop to use `vse8`.

Doesn't look like alignment was accounted for in the intrinsic definition. masked.gather/scatter had a separate argument. vp.gather/scatter use a pointer attribute. I think we should use the pointer attribute for compress store.

It looks like SelectionDAGBuilder::visitMaskedStore currently sets the alignment to the ABI alignment of the vector type which doesn't make sense.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1613]: Once we fix the alignment for the intrinsic we'll also need to teach ScalarMaskedMemIntrinsic to preserve it when scalarizing.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -1609,3 +1609,10 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
...
+  auto *VTy = dyn_cast<VectorType>(DataTy);
+  if (!VTy || VTy->isScalableTy() || !ST->hasVInstructions())
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1615]: Need to check that the element type is supported.

File: llvm/test/CodeGen/RISCV/rvv/compressstore.ll (hunk 1 above)
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/compressstore.ll:17540]: Drop this

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 2 above)
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: > Once we fix the alignment for the intrinsic we'll also need to teach ScalarMaskedMemIntrinsic to preserve it when scalarizing.

To emphasize, this should be done in a separate review and should probably be addressed before we continue too far with this one.

**Comment by preames** [comment]
Can you separate your NFC to add the Align parameter to TTI into it's own review?  It's clearly needed, we should be able to land that one quickly and rebase this on top.

**Comment by nikolaypanchenko** [comment]
> Can you separate your NFC to add the Align parameter to TTI into it's own review? It's clearly needed, we should be able to land that one quickly and rebase this on top.

Absolutely. https://github.com/llvm/llvm-project/pull/83516

File: llvm/test/CodeGen/RISCV/rvv/compressstore.ll [hunk 4]
```diff
@@ -0,0 +1,15266 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -verify-machineinstrs -mtriple=riscv64 -mattr=+v,+d %s -o - | FileCheck %s --check-prefix=RV64
+; RUN: llc -verify-machineinstrs -mtriple=riscv32 -mattr=+v,+d %s -o - | FileCheck %s --check-prefix=RV32
+
+; Compress + store for i8 type
+
+define void @test_compresstore_i8_v1(ptr %p, <1 x i1> %mask, <1 x i8> %data) {
```
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/compressstore.ll:7]: Nit usually the vectors in a test name use the MVT naming, e.g. `v1i8`/`nxv2i32`

File: llvm/test/CodeGen/RISCV/rvv/compressstore.ll
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/compressstore.ll:1]: Is it possible to test for scalable vectors too?

Comment (nikolaypanchenko) [reply llvm/test/CodeGen/RISCV/rvv/compressstore.ll:1]: TTI part of the changeset disallows scalable vectors. The legalization part in `ScalarizeMaskedMemIntrin` cannot handle them either.
It might be better to relax TTI constraint and add tests for scalable vectors in followup patch, what do you think ?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 5]
```diff
@@ -1620,3 +1620,15 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
                   C2.NumIVMuls, C2.NumBaseAdds,
                   C2.ScaleCost, C2.ImmCost, C2.SetupCost);
 }
+
+bool RISCVTTIImpl::isLegalMaskedCompressStore(Type *DataTy, Align Alignment) {
```
[the remark this answers is not in the document]
Comment (nikolaypanchenko) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1624]: Unlike x86, if store is done via separate instruction and requires masking, RVV's implementation does not require masking/predication due to tweaked VL.

[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1624]: The check being asked for is that fixed vectors are supported. Not that the specific type is legal. We still have a master disable for fixed vectors in the backend by setting -rvv-vector-bits-min=0.

```
    // Only support fixed vectors if we know the minimum vector size.
    if (DataTypeVT.isFixedLengthVector() && !ST->useRVVForFixedLengthVectors())
      return false;
```

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 6]
```diff
@@ -1620,3 +1620,15 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
...
+  auto *VTy = dyn_cast<VectorType>(DataTy);
+  if (!VTy || VTy->isScalableTy() || !ST->hasVInstructions())
+    return false;
+
+  if (!TLI->isLegalElementTypeForRVV(
+          TLI->getValueType(DL, VTy->getElementType())))
+    return false;
+
+  return getRegUsageForType(VTy) <= 8;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1633]: Why is this needed? The backend should be able to split a compressed store.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 7]
```diff
@@ -1620,3 +1620,15 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
...
+
+bool RISCVTTIImpl::isLegalMaskedCompressStore(Type *DataTy, Align Alignment) {
+  auto *VTy = dyn_cast<VectorType>(DataTy);
+  if (!VTy || VTy->isScalableTy() || !ST->hasVInstructions())
+    return false;
+
+  if (!TLI->isLegalElementTypeForRVV(
+          TLI->getValueType(DL, VTy->getElementType())))
+    return false;
+
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1632]: Need to check alignment unless you add misaligned support to RISCVISelLowering.cpp

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 6 above)
Comment (nikolaypanchenko) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1633]: It's not for the store, but for the compress. Splitting it for LMUL >= 16 is doable, but is not simple

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 7 above)
Comment (nikolaypanchenko) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1632]: Ok. I'll add `isLegalMaskedLoadStore` as @yetingk suggested.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 6 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1637]: The splitting will happen while it is still a compressedstore. The compress instruction isn't created until after the split is already done.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 8]
```diff
@@ -1620,3 +1620,19 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
                   C2.NumIVMuls, C2.NumBaseAdds,
                   C2.ScaleCost, C2.ImmCost, C2.SetupCost);
 }
+
+bool RISCVTTIImpl::isLegalMaskedCompressStore(Type *DataTy, Align Alignment) {
+  auto *VTy = dyn_cast<VectorType>(DataTy);
+  if (!VTy || VTy->isScalableTy() || !ST->hasVInstructions())
+    return false;
+
+  if (!TLI->isLegalElementTypeForRVV(
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1629]: This is redundant with the call to `isLegalMaskedLoadStore`

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 9]
```diff
@@ -1620,3 +1620,19 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
                   C2.NumIVMuls, C2.NumBaseAdds,
                   C2.ScaleCost, C2.ImmCost, C2.SetupCost);
 }
+
+bool RISCVTTIImpl::isLegalMaskedCompressStore(Type *DataTy, Align Alignment) {
+  auto *VTy = dyn_cast<VectorType>(DataTy);
+  if (!VTy || VTy->isScalableTy() || !ST->hasVInstructions())
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1626]: The hasVInstructions check is redundant with `isLegalMaskedLoadStore`

**Review by topperc — APPROVED** [review]
LGTM
